File: 02 February/02-17/checkEligibility.js
function checkEligibility(athleteWeights, sledWeight) {
  const onePersonBobsledMinWeight = 162;
  const twoPersonBobsledMinWeight = 170;
  const fourPersonBobsledMinWeight = 210;
  const onePersonMaxTotalWeight = 247;
  const twoPersonMaxTotalWeight = 390;
  const fourPersonMaxTotalWeight = 630;

  const totalWeight = athleteWeights.reduce((total, weight) => total + weight) + sledWeight;

  switch (athleteWeights.length) {
    case 1:
      if (sledWeight < onePersonBobsledMinWeight || totalWeight > onePersonMaxTotalWeight)
        return "Not Eligible";
      break;
    case 2:
      if (sledWeight < twoPersonBobsledMinWeight || totalWeight > twoPersonMaxTotalWeight)
        return "Not Eligible";
      break;
    case 4:
      if (sledWeight < fourPersonBobsledMinWeight || totalWeight > fourPersonMaxTotalWeight)
        return "Not Eligible";
  }

  return "Eligible";
}

// --- TEST SUITE ---

const testsText = String.raw`
1. checkEligibility([78], 165) should return "Eligible".
2. checkEligibility([80], 160) should return "Not Eligible".
3. checkEligibility([80], 170) should return "Not Eligible".
4. checkEligibility([85, 90], 170) should return "Eligible".
5. checkEligibility([85, 95], 168) should return "Not Eligible".
6. checkEligibility([112, 97], 185) should return "Not Eligible".
7. checkEligibility([110, 102, 90, 106], 222) should return "Eligible".
8. checkEligibility([106, 99, 90, 88], 205) should return "Not Eligible".
9. checkEligibility([106, 99, 103, 96], 227) should return "Not Eligible".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 02 February/02-17/check_eligibility.py
def check_eligibility(athlete_weights, sled_weight):
    one_person_bobsled_min_weight = 162
    two_person_bobsled_min_weight = 170
    four_person_bobsled_min_weight = 210
    one_person_max_total_weight = 247
    two_person_max_total_weight = 390
    four_person_max_total_weight = 630

    total_weight = sum(athlete_weights) + sled_weight

    match len(athlete_weights):
        case 1:
            if sled_weight < one_person_bobsled_min_weight or total_weight > one_person_max_total_weight:
                return "Not Eligible"
        case 2:
            if sled_weight < two_person_bobsled_min_weight or total_weight > two_person_max_total_weight:
                return "Not Eligible"
        case 4:
            if sled_weight < four_person_bobsled_min_weight or total_weight > four_person_max_total_weight:
                return "Not Eligible"

    return "Eligible"

# --- TEST SUITE ---

tests_text = r'''
1. check_eligibility([78], 165) should return "Eligible".
2. check_eligibility([80], 160) should return "Not Eligible".
3. check_eligibility([80], 170) should return "Not Eligible".
4. check_eligibility([85, 90], 170) should return "Eligible".
5. check_eligibility([85, 95], 168) should return "Not Eligible".
6. check_eligibility([112, 97], 185) should return "Not Eligible".
7. check_eligibility([110, 102, 90, 106], 222) should return "Eligible".
8. check_eligibility([106, 99, 90, 88], 205) should return "Not Eligible".
9. check_eligibility([106, 99, 103, 96], 227) should return "Not Eligible".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)